### ISSUE TITLE
Discuss and document test helper style guidelines

### ISSUE THREAD

**Issue by p0mvn** [osmosis-labs/osmosis#6520]
## Background

There are various ways to define helpers in tests.

**Approach A:**
- Define them in the test-global scope

Pros:
- More readable

Cons:
- Bloats test-global scope
Example:
https://github.com/osmosis-labs/osmosis/blob/850b0415ff4ed996b1491fc58debe4245c12c766/x/incentives/keeper/keeper_test.go#L36

**Approach B:**
- Define them in the restricted per-test scope

Pros:
- Restricts helpers to only be exposed to the scope where they are useful and reduces global scope bloat
Cons:
- Less readable

Example:
https://github.com/osmosis-labs/osmosis/blob/850b0415ff4ed996b1491fc58debe4245c12c766/x/incentives/keeper/distribute_test.go#L2110-L2134

We should discuss and converge on the preferred approach to follow universally as a team.

## Acceptance Criteria
- One of the approaches is selected
- Documented in CONTRIBUTING.md
- Used universally in PRs
- Consider implementing linter to prefer the desired style

**Comment by p0mvn** [comment]
I have preferred utilizing Approach A whenever the function is only used in one test. If more tests need to reuse the helper, I would convert the helper to Approach B.

**Comment by AlpinYukseloglu** [comment]
One alternative that we have implicitly done in recent modules e.g. CL is to set scope for helpers by file. For instance, a helper that only applies to incentive logic would be in the `incentives_test.go` file, and would be defined globally even if at the time it is written it is only used for that test (which will be the case for pretty much all useful helpers when they are introduced).

This approach has three immediate benefits:
1. It avoids having tens (or at times 100+) lines of setup code of the form `functionA := func { ... }` which I think is a huge readability win
2. It minimizes future test refactoring overhead and/or unintentional duplicate code if it ever becomes the case that the helpers defined in test scope are useful elsewhere. This also means we lean less on developers being able to (and taking the time to) predict future categories of tests that will be written.
3. (Less clear of a win) It keeps test helpers consistent with our higher level test suite (e.g. how general tests might be defined in `keeper_test.go` and even broader ones in apptesting)

The primary con is that we risk ending up with test files that have a large section at the beginning or end dedicated to helpers and constants/default values. I believe that this is still a net test readability (and thus overall effectiveness) win because it keeps the actual tests themselves much more lightweight and makes it straightforward to scan a test and see what it tested/asserted (especially if we have effective names for the helpers).

**Comment by p0mvn** [comment]
I agree with all the benefits listed and leaning towards agreeing with the proposed approach. 

However, for completeness of the trade-off space analysis, another con to consider is that a helper that is defined in file A (e.g. `incentives_test.go`) is still accessible in another file B (e.g. spread_factor_test.go). As a result, if the developer begins writing spread factor tests, their autocomplete will be polluted with all other incentive-specific helpers in the global scope.

> The primary con is that we risk ending up with test files that have a large section at the beginning or end dedicated to helpers and constants/default values. 

Thoughts on defining helpers at the bottom of the file instead of at the top? Whenever a developer opens a new test file, it might be more desirable to start reviewing the higher-level test fixtures as opposed to low-level test helpers. I always find myself scrolling/searching to skip helpers whenever we have helpers at the top. As a result, I think having them at the bottom might be more streamlined long term.

**Comment by AlpinYukseloglu** [comment]
> However, for completeness of the trade-off space analysis, another con to consider is that a helper that is defined in file A (e.g. incentives_test.go) is still accessible in another file B (e.g. spread_factor_test.go). As a result, if the developer begins writing spread factor tests, their autocomplete will be polluted with all other incentive-specific helpers in the global scope.

This is a good point that I had not considered, especially if this type of clutter matters to other folks on the team (it personally does not make that much of a difference for me unless there's heavy name conflicts between helpers). A possible solution (which might be a generally useful thing to do) might be to prefix helpers to categorize them if we're able to come up with a pattern of prefixes that isn't overly verbose.

> Thoughts on defining helpers at the bottom of the file instead of at the top? Whenever a developer opens a new test file, it might be more desirable to start reviewing the higher-level test fixtures as opposed to low-level test helpers. I always find myself scrolling/searching to skip helpers whenever we have them at the top. As a result, I think having them at the bottom might be more streamlined long term.

This was generally what I had in mind as well. One important design goal for tests imo is to be able to open a test file and scan to understand coverage, and having helpers be either 1. at the top or 2. scattered in between gets in the way of this, but as long as we're consistent I'd be happy with whichever we land on.

**Comment by PaddyMc** [comment]
Approach A is more readable IMO, but it really depends on what we're optimising for, I happy with either, but if it's a 50/50 decision the best option is to go with what's most readable and understandable.
